### DISCUSSION TITLE
Some general questions about reticulum/nomadnetwork

### DISCUSSION THREAD

**Discussion by silveroranges** [markqvist/Reticulum#63] [General]
Hello again! I have been playing around with reticulum and nomadnetwork, along with the utilities that reticulum comes with. I also read the manual back to back. Here are my questions so far about how the network functions;

On page 48, in the link establishment section. What happens if a node in between the two nodes that created a link drops out? Is a new link renegotiated, or is a route around the dropped out node found?

Also, on page 49, also handling links, the 4th bullet point down says the node 'will decide whether to accept the request', what variables drive the decision to accept/reject the request packet?

Page 39, 'the interface facing the clients must be put into gateway mode', just to make sure I understand this correctly. If I have a wireless router for my home, and a rpi attached to the network via wifi on one interface, and also a LoRa interface, the interface I put in gateway mode is going to be the wireless interface right? Since it is what will be serving the clients that connect via sideband or nomadnet.

Also I seem to be having problems with announce forwarding. For example, as a test I have two raspberry pi... actually I will upload a drawing because it will be easier to draw than explain the topology.
![image](https://user-images.githubusercontent.com/22848776/172013708-f99ddf47-74a2-4253-a499-17c7dedb1ccd.png)

The rpi_1 and rpi_2 are able to communicate with each other, as well as rpi_1 and phone. But the announces are not traveling to the rpi_2 from the phone, or vice versa. I put the interface facing the w_lan into gateway mode on rpi_1. I tried manually adding the address of rpi_2 and requesting its keys from the network but that failed as well. 

I am very excited about getting a good reticulum network up and running, I am in the phase of trying to understand it all in depth so that I can explain it to other people, I want to create an easy to make node that just requires a few files, raspberry pi/lora and a 3d printed case. 

Also on that matter, would it be possible to run a node on a raspberry pi pico? It can run micropython, and seems to have enough memory. I was looking for a few raspberry pi 4's or 3b's, and they are still stupidly overpriced if they are even available. I have like 30 picos that I was using for another project though.

**Comment by 4c3e** [comment]
Hello! 
> On page 48, in the link establishment section. What happens if a node in between the two nodes that created a link drops out? Is a new link renegotiated, or is a route around the dropped out node found?

Pretty sure the link will time out and an error would be thrown. It would be up to the application to request a new link.

> Also, on page 49, also handling links, the 4th bullet point down says the node 'will decide whether to accept the request', what variables drive the decision to accept/reject the request packet?

This can be whatever the application chooses. See the RNS Destination API (specifically the "register_request_handler"): https://markqvist.github.io/Reticulum/manual/reference.html#destination
A common use is to only accept requests from whitelisted identities for authentication purposes.

>Page 39, 'the interface facing the clients must be put into gateway mode', just to make sure I understand this correctly. If I have a wireless router for my home, and a rpi attached to the network via wifi on one interface, and also a LoRa interface, the interface I put in gateway mode is going to be the wireless interface right? Since it is what will be serving the clients that connect via sideband or nomadnet.

Yes, that's how I understand it as well.

>The rpi_1 and rpi_2 are able to communicate with each other, as well as rpi_1 and phone. But the announces are not traveling to the rpi_2 from the phone, or vice versa. I put the interface facing the w_lan into gateway mode on rpi_1. I tried manually adding the address of rpi_2 and requesting its keys from the network but that failed as well.

Hmm. I'm not sure what the problem could be. Are you still experiencing this issue? The announcements should be propagating like you would expect.

> I am very excited about getting a good reticulum network up and running, I am in the phase of trying to understand it all in depth so that I can explain it to other people, I want to create an easy to make node that just requires a few files, raspberry pi/lora and a 3d printed case.

Awesome! 

> Also on that matter, would it be possible to run a node on a raspberry pi pico? It can run micropython, and seems to have enough memory. I was looking for a few raspberry pi 4's or 3b's, and they are still stupidly overpriced if they are even available. I have like 30 picos that I was using for another project though.

I'm not sure how similar the hardware is, but a member of the community put together a RNS node guide using a raspi zero: https://github.com/greygoo/Mesh_box

Might be a place to start?

**Comment by markqvist** [comment]
Hi @silveroranges and @4c3e! I want to add in a few confirmations and extra points here as well.

> But the announces are not traveling to the rpi_2 from the phone, or vice versa.

Did you enable transport on rpi_1? If not, in the config, set `enable_transport = yes`. Announce propagation and traffic routing is only handled by transport instances.

> Page 39, 'the interface facing the clients must be put into gateway mode', just to make sure I understand this correctly. If I have a wireless router for my home, and a rpi attached to the network via wifi on one interface, and also a LoRa interface, the interface I put in gateway mode is going to be the wireless interface right?

Yes, this is correct.

> I'm not sure how similar the hardware is, but a member of the community put together a RNS node guide using a raspi zero: https://github.com/greygoo/Mesh_box

I think that it will be a very good place to start. I have not tried the build yet myself, since I don't have the necessary hardware yet. But I know @greygoo has put a lot of effort into it, and it should be working well. I think he was planning on having images with RNS pre-installed, but for now you just have to manually flash a normal armbian image and install RNS via pip.

> Also on that matter, would it be possible to run a node on a raspberry pi pico?

Expect some announcements in those regards very soon :) Recent commits to the Reticulum repo might offer some hints ;) But to not cause any unnecessary expectations, I will say that, initially at least, the ability to run a transport instance on microcontrollers will be disabled (until we have explored how well the basic stack, without transport, runs on these kinds of very small systems).

Running a transport instance does require quite a bit of memory and processing power. Of course it is all relative to the amount of traffic and destinations. We might be able to pull it off on a modern MCU like the RP2040 or ESP32 with an SD card to make up for the memory reuqirements. It's on the horizon, but we still have some way to go there :)

> Also, on page 49, also handling links, the 4th bullet point down says the node 'will decide whether to accept the request', what variables drive the decision to accept/reject the request packet?

Hmm, that section is confusing. There is two problems here, one being I didn't write that section concisely, the other being that a piece of functionality is missing in the API. Thanks for pointing it out, I will fix it shortly. At the time of link establishment and key exchange, there is no information available on who the link initiator is, and the only decision that can be made is a binary one: *whether the destination accepts links at all*. This setting is missing in the API, and I will fix that shortly, along with updating the wording in the manual.

Once a link has been established, the initiator has the option of *identifying* towards the destination, and decisions can be made on the basis of that, you can see [this part of the rncp utility](https://github.com/markqvist/Reticulum/blob/master/RNS/Utilities/rncp.py#L98) for an example. There is also the API options for [arbitrary checks on requests](https://markqvist.github.io/Reticulum/manual/reference.html#RNS.Destination.register_request_handler) and [resource advertisements](https://markqvist.github.io/Reticulum/manual/reference.html#RNS.Destination.register_request_handler) over established links.

Hope that clears it up a bit, I will rework that part of the docs to make it more clear, and add the missing APIs shortly.

**Reply by greygoo** [reply]
Hi guys, was off for a while, currently updating the armbian build system and finishing the documentation,  sorry for the delay.
